### PR TITLE
System prompt

### PR DESCRIPTION
**Pull request by psinger**
Closes https://github.com/h2oai/h2o-llmstudio/issues/223

Currently not working for rlhf

Tooltips will follow after final decisions are made

### CODE REVIEW

File: llm_studio/src/datasets/text_causal_language_modeling_ds.py [hunk 1]
```diff
@@ -304,23 +335,32 @@ def sanity_check(cls, df: pd.DataFrame, cfg: Any, mode: str = "train"):
...
         encodings = parent_encodings + encodings
 
-        sample["reward_model_prompt_text"] = (
-            reward_model_parent_prompt_text + self.raw_prompts[idx]
-        )
+        # get first system encoding
+        system_encoding = encodings[0][0]
+        # remove system encodings from list of encodings
+        encodings = [encoding[1:] for encoding in encodings]
+        encodings[0][0] = torch.cat([system_encoding, encodings[0][0]])
```
Comment (pascal-pfeiffer) [inline llm_studio/src/datasets/text_causal_language_modeling_ds.py:358]: I don't fully grasp that part and why we append it already here. Should it be prepended at the end maybe?

File: llm_studio/src/possible_values.py [hunk 2]
```diff
@@ -25,7 +25,7 @@ def _scan_dirs(dirname) -> List[str]:
 
 
 def _scan_files(
-    dirname, extensions: Tuple[str, str, str] = (".csv", ".pq", ".json")
+    dirname, extensions: Tuple[str, ...] = (".csv", ".pq", ".parquet", ".json")
```
Comment (pascal-pfeiffer) [inline llm_studio/src/possible_values.py:28]: :+1:

**Review by pascal-pfeiffer — CHANGES_REQUESTED** [review]
Thanks a lot for the functionality @psinger 
- I would suggest that we also add it to the validation visualizations. 
- Can we add it to the chat interface somehow?
- as you said, tooltips are missing

I have a bit of understanding issues why we append the systems prompt embedding as early. Probably a bit tricky with the chained conversations.

File: llm_studio/src/datasets/text_causal_language_modeling_ds.py (hunk 1 above)
Comment (psinger) [reply llm_studio/src/datasets/text_causal_language_modeling_ds.py:358]: So currently each sample in `encodings` has the system prompt, but we only need it for the root, which is why we pick it from `encodings[0][0]` and then remove it again, so that we only have prompt/answer as before.
Then we just concat it to the very first prompt and have the typical workflow.

Then we only need to have this extra functionality at the end, to make sure it is prepended if max token length is exceeded. Otherwise we would always need to prepend it manually, which we could also do, but a bit more messy because we would need to do it after the padding, and then we would need to find the index where the padding ends etc.

Bit messy, but "simplest" solution I could think of. Im sure there are better ways.

**Comment by psinger** [comment]
> Can we add it to the chat interface somehow?

We could, but do you think it would be an important feature? would need to be an extra input then

> I would suggest that we also add it to the validation visualizations.

Lets add it here: https://github.com/h2oai/h2o-llmstudio/issues/255

File: llm_studio/src/datasets/text_causal_language_modeling_ds.py (hunk 1 above)
Comment (pascal-pfeiffer) [reply llm_studio/src/datasets/text_causal_language_modeling_ds.py:358]: Thank you, now I understand. Makes perfect sense.
Could you maybe add a few short comments in the code for that reasoning. Basically what you said above.

**Comment by pascal-pfeiffer** [comment]
> > Can we add it to the chat interface somehow?
> 
> We could, but do you think it would be an important feature? would need to be an extra input then
> 
> > I would suggest that we also add it to the validation visualizations.
> 
> Lets add it here: #255


Agree, this can definitely be done in follow-up issues.
Else, looks good to me, besides the missing tooltips.
